### PR TITLE
add compatibility with pathlib.Path to root_io

### PR DESCRIPTION
**Pull request by Victor-Schwan**
BEGINRELEASENOTES
- add compatibility with pathlib.Path objects in root_io and sio_io

ENDRELEASENOTES

### CODE REVIEW

**Review by tmadlener — COMMENTED** [review]
I suppose you stumbled over this when trying to instantiate a `Reader` with a `pathlib.Path` object and it failed?

The same issue will be present in `sio_io`, so it should also be fixed there.

File: python/podio/root_io.py [hunk 1]
```diff
@@ -10,18 +11,39 @@
...
+    if isinstance(filenames, list):
+        str_filenames = []
+        for f in filenames:
+            if isinstance(f, (str, Path)):
+                str_filenames.append(str(f))
+            else:
+                raise TypeError(f"Invalid filename type: {f} (type: {type(f)})")
+        return str_filenames
+
+    raise TypeError(f"Invalid filenames argument: {filenames} (type: {type(filenames)})")
```
[the remark this answers is not in the document]
Comment (Victor-Schwan) [reply python/podio/root_io.py:34]: Ok. Do you have any objections to using `os.fspath` instead of `str`? It is better for catching wrong types, e.g. if `None` is passed

[the remark this answers is not in the document]
Comment (tmadlener) [reply python/podio/root_io.py:34]: If `reader.openFiles` swallows a list of `os.fspath`, I have no objections.

[the remark this answers is not in the document]
Comment (tmadlener) [reply python/podio/root_io.py:34]: If `reader.openFiles` swallows a list of `os.fspath`, I have no objections.

[the remark this answers is not in the document]
Comment (Victor-Schwan) [reply python/podio/root_io.py:34]: `os.fspath()` only returns `str` or `bytes`. I am unsure what happens if a `bytes` path is given to `reader.openFiles`. `fspath` is just better at throwing `TypeError` if the argument is not `pathlike` compared to `str()`. https://docs.python.org/3/library/os.html#os.fspath

File: python/podio/sio_io.py [hunk 2]
```diff
@@ -20,8 +21,9 @@ def __init__(self, filename):
         """Create a reader that reads from the passed file.
 
         Args:
-            filename (str): File to open and read data from
+            filename (str or Path): File to open and read data from.
         """
+        filename = convert_to_str_paths(filename)
```
Comment (tmadlener) [inline python/podio/sio_io.py:26]:
```suggestion
        filename = convert_to_str_paths(filename)[0]
```
As `openFile` takes a single string.

File: python/podio/sio_io.py [hunk 3]
```diff
@@ -49,13 +52,15 @@ def __init__(self, filename):
...
 
     def __init__(self, filename):
-        """Create a writer for writing files
+        """Create a writer for writing files.
 
         Args:
-            filename (str): The name of the output file
+            filename (str or Path): The name of the output file.
         """
+        filename = convert_to_str_paths(filename)
```
Comment (tmadlener) [inline python/podio/sio_io.py:63]:
```suggestion
        filename = convert_to_str_paths(filename)[0]
```

File: python/podio/root_io.py [hunk 4]
```diff
@@ -10,18 +13,34 @@
...
+    Returns:
+        list[str]: A list of filepaths as strings.
+    """
+
+    if isinstance(filenames, Iterable) and not isinstance(filenames, (str, Path)):
+        return [os.fspath(fn) for fn in filenames]
+
+    return [os.fspath(filenames)]
+
+
```
Comment (tmadlener) [inline python/podio/root_io.py:33]: I would move this into a separate module that can then be imported by `root_io` and `sio_io` instead of cross-importing this from `root_io` to `sio_io`. Altough I generally don't like `utils` too much, I can't think of a better name at the moment, so I would be OK with that.

File: python/podio/root_io.py [hunk 5]
```diff
@@ -3,11 +3,17 @@
...
 from ROOT import podio  # noqa: E402 # pylint: disable=wrong-import-position
 
-from podio.base_reader import BaseReaderMixin  # pylint: disable=wrong-import-position # noqa: E402
-from podio.base_writer import BaseWriterMixin  # pylint: disable=wrong-import-position # noqa: E402
+from podio.base_reader import (  # pylint: disable=wrong-import-position # noqa: E402
+    BaseReaderMixin,
+)
+from podio.base_writer import (  # pylint: disable=wrong-import-position # noqa: E402
+    BaseWriterMixin,
+)
```
Comment (tmadlener) [inline python/podio/root_io.py:16]:
```suggestion
from podio.base_reader import BaseReaderMixin  # pylint: disable=wrong-import-position # noqa: E402
from podio.base_writer import BaseWriterMixin  # pylint: disable=wrong-import-position # noqa: E402
```
No unrelated format changes please

File: python/podio/utils.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import os
```
Comment (tmadlener) [inline python/podio/utils.py:1]:
```suggestion
#!/usr/bin/env python3
"""utility functionality for podio"""

import os
```
